File: android/app/src/main/java/com/tb/fkst/core/Signer.kt
package com.tb.fkst.core

import java.security.MessageDigest

/**
 * 签名算法（三种变体），与官方 APP 完全一致。
 *
 * 通用：  MD5( 字典序拼接 KV + SALT + MD5(tpl)[5:21] ).upper()
 *   其中 tpl = "f0{call_id 后四位}com.yaerxing.fkst{app_c}F.K*$t"
 *
 * 文章详情：把 call_id 换成 timestamp，app_c 换成 app_v
 * 发评论：  只签 api_key + call_id + openid 三个字段
 */
object Signer {

    private const val TEXT_TPL = "f0%scom.yaerxing.fkst%sF.K*\$t"
    private const val NOTE_TPL = "f0%s%sF.K*\$t"

    fun md5(input: String): String {
        val md = MessageDigest.getInstance("MD5")
        val bytes = md.digest(input.toByteArray(Charsets.UTF_8))
        val sb = StringBuilder(bytes.size * 2)
        for (b in bytes) {
            val v = b.toInt() and 0xFF
            if (v < 0x10) sb.append('0')
            sb.append(Integer.toHexString(v))
        }
        return sb.toString()
    }

    private fun tail4(value: String?): String {
        val v = value ?: return ""
        return if (v.length <= 4) v else v.substring(v.length - 4)
    }

    private fun digestOf(tpl: String): String = md5(tpl).substring(5, 21)

    /** 通用签名 */
    fun signParams(params: Map<String, String>, appC: String = "171"): String {
        val body = StringBuilder()
        params.entries
            .filter { it.key != "api_sig" }
            .sortedBy { it.key }
            .forEach { body.append(it.key).append(it.value) }
        body.append(Constants.SIGN_SECRET)
        body.append(digestOf(TEXT_TPL.format(tail4(params["call_id"]), appC)))
        return md5(body.toString()).uppercase()
    }

    /** H5 文章详情专用：用 timestamp 后四位 */
    fun signNote(params: Map<String, String>): String {
        val body = StringBuilder()
        params.entries
            .filter { it.key != "api_sig" }
            .sortedBy { it.key }
            .forEach { body.append(it.key).append(it.value) }
        body.append(Constants.SIGN_SECRET)
        val appV = Constants.GET_NOTE_PARAMS["app_v"] ?: "171"
        body.append(digestOf(NOTE_TPL.format(tail4(params["timestamp"]), appV)))
        return md5(body.toString()).uppercase()
    }

    /** 发评论专用 */
    fun signComment(params: Map<String, String>): String {
        val callId = params["call_id"] ?: error("评论签名缺少参数: call_id")
        val apiKey = params["api_key"] ?: error("评论签名缺少参数: api_key")
        val openid = params["openid"] ?: error("评论签名缺少参数: openid")
        val appC = params["app_c"] ?: "171"
        val body = StringBuilder()
            .append("api_key").append(apiKey)
            .append("call_id").append(callId)
            .append("openid").append(openid)
            .append(Constants.SIGN_SECRET)
            .append(digestOf(TEXT_TPL.format(tail4(callId), appC)))
        return md5(body.toString()).uppercase()
    }

    fun sign(kind: String, params: Map<String, String>): String = when (kind) {
        "note" -> signNote(params)
        "comment" -> signComment(params)
        else -> signParams(params)
    }

    /** 登录密码 = MD5(明文 + 盐) */
    fun encryptPassword(password: String, salt: String = Constants.PWD_SALT): String =
        md5(password + salt)
}

File: fkst_sdk/signer.py
"""签名算法（三种变体），与官方 APP 一致。"""

import hashlib

from .constants import API_KEY, SIGN_SECRET, GET_NOTE_PARAMS

# 官方模板串： f0{call_id 后四位}com.yaerxing.fkst{app_c}F.K*$t
_TEXT_TPL = "f0{cid}com.yaerxing.fkst{app_c}F.K*$t"
# H5 文章详情模板串：f0{timestamp 后四位}{app_v}F.K*$t
_TEXT_TPL_NOTE = "f0{ts}{app_v}F.K*$t"


def _md5(s: str) -> str:
    return hashlib.md5(s.encode("utf-8")).hexdigest()


def _stamp(params: dict, key: str, tpl: str, **kw) -> str:
    """结果串 = 字典序拼接 KV + SECRET + md5(模板)[5:21]"""
    body = "".join(
        f"{k}{v}" for k, v in sorted(params.items()) if k != "api_sig"
    ) + SIGN_SECRET
    suffix = str(params[key])[-4:]
    return body + _md5(tpl.format(cid=suffix, ts=suffix, **kw))[5:21]


def sign_params(params: dict, app_c: str = "171") -> str:
    """通用签名。"""
    return _md5(_stamp(params, "call_id", _TEXT_TPL, app_c=app_c)).upper()


def sign_note(params: dict) -> str:
    """GET_NOTE 专用：用 timestamp 后四位，不含 call_id。"""
    return _md5(
        _stamp(params, "timestamp", _TEXT_TPL_NOTE, app_v=GET_NOTE_PARAMS["app_v"])
    ).upper()


def sign_comment(params: dict) -> str:
    """发评论专用：只签 api_key + call_id + openid 三个字段。"""
    for k in ("call_id", "api_key", "openid"):
        if k not in params:
            raise ValueError(f"评论签名缺少参数: {k}")
    body = (
        f"api_key{params['api_key']}call_id{params['call_id']}openid{params['openid']}"
        + SIGN_SECRET
    )
    body += _md5(
        _TEXT_TPL.format(cid=str(params["call_id"])[-4:], app_c=params.get("app_c", "171"))
    )[5:21]
    return _md5(body).upper()


def encrypt_password(password: str, salt: str) -> str:
    """登录密码：md5(明文 + 盐)。"""
    return _md5(password + salt)


SIGNERS = {
    "default": sign_params,
    "note": sign_note,
    "comment": sign_comment,
}
